app/components/file_attachment_list/file_attachment.js

```diff
@@ -28,7 +49,128 @@ export default class FileAttachment extends PureComponent {
         onInfoPress: () => true,
         onPreviewPress: () => true
-    }
+    };
+
+    static contextTypes = {
+        intl: intlShape
+    };
+
+    state = {
+        didCancel: false,
+        downloading: false,
+        progress: 0
+    };
+
+    cancelDownload = () => {
+        this.setState({didCancel: true});
+        if (this.downloadTask) {
+            this.downloadTask.cancel();
+        }
+    };
+
+    downloadAndPreviewFile = async (file) => {
+        const path = `${DOCUMENTS_PATH}/${file.name}`;
+
+        try {
+            const isDir = await RNFetchBlob.fs.isDir(DOCUMENTS_PATH);
+            if (!isDir) {
+                try {
+                    await RNFetchBlob.fs.mkdir(DOCUMENTS_PATH);
+                } catch (error) {
+                    // Do nothing
+                }
+            }
+
+            const options = {
+                session: file.id,
+                timeout: 10000,
+                indicator: true,
+                overwrite: true,
+                path
+            };
+
+            const exist = await RNFetchBlob.fs.exists(path);
+            if (exist) {
+                this.openDocument(file, 0);
+            } else {
+                this.setState({downloading: true});
+                this.downloadTask = RNFetchBlob.config(options).fetch('GET', Utils.getFileUrl(file.id)).
+                    progress((received, total) => {
+                        const progress = (received / total) * 100;
+                        this.setState({progress});
+                    });
+
+                await this.downloadTask;
+                this.setState({
+                    progress: 100
+                }, () => {
+                    // need to wait a bit for the progress circle UI to update to the give progress
+                    this.openDocument(file);
+                });
+            }
+        } catch (error) {
+            RNFetchBlob.fs.unlink(path);
+            this.setState({downloading: false, progress: 0});
+
+            if (error.message !== 'cancelled') {
+                const {intl} = this.context;
+                Alert.alert(
+                    intl.formatMessage({
+                        id: 'mobile.downloader.failed_title',
+                        defaultMessage: 'Download failed'
+                    }),
+                    intl.formatMessage({
+                        id: 'mobile.downloader.failed_description',
+                        defaultMessage: 'An error occurred while downloading the file. Please check your internet connection and try again.\n'
+                    }),
+                    [{
+                        text: intl.formatMessage({
+                            id: 'mobile.server_upgrade.button',
+                            defaultMessage: 'OK'
+                        })
+                    }]
+                );
+            }
+        }
+    };
+
+    openDocument = (file, delay = 2000) => {
+        setTimeout(() => {
```

Comment:
Might be a good idea to check if the component is still mounted somewhere in here. The case that I'm thinking of is that someone starts a download and then switches channels.